### ISSUE TITLE
Avoid compacting chunks that is already compacted

### ISSUE THREAD

**Issue by NGA-TRAN** [influxdata/influxdb_iox#1887]
This is a sub-task of #1885 and also described in [this document](https://docs.google.com/document/d/1xn1zxxDNZ6ufjZK-texoaBjj5dP5irfbuk2WoWXgWxQ/edit#). The problem is we keep compacting already compacted RUB data with new MUB which means our compacted RUB gets larger and larger and we redundantly repeat running the expensive data scan.

 ```
                value of time column in data                                                                        
   │            ───────────────────────────────────────────────────────────────────────    ...   ─────────────────▶   
   │                                                                                                                  
   │  Initial    ┌ ─ ─ ─ ─ ─ ─                                                                                        
   │   State          MUB     │                                                                                       
   │             └ ─ ─ ─ ─ ─ ─                                                                                        
   │                                                                                                                  
   │             ┌────────────┐┌ ─ ─ ─ ─ ─ ─                                                                          
   │  Compaction │    RUB     │     MUB     │                                                                         
   │      1      └────────────┘└ ─ ─ ─ ─ ─ ─                                                                          
   │                                                                                                                  
   │             ┌─────────────────────────┐┌ ─ ─ ─ ─ ─ ─                                                             
   │  Compaction │           RUB           │     MUB     │                                                            
   │      2      └─────────────────────────┘└ ─ ─ ─ ─ ─ ─                                                             
   │                                                                                                                  
   │             ┌───────────────────────────────────────┐┌ ─ ─ ─ ─ ─ ─                                               
   │  Compaction │                  RUB                  │     MUB     │                                              
   │      3      └───────────────────────────────────────┘└ ─ ─ ─ ─ ─ ─                                               
   │                                                                                                                  
   │      ...                                                                                                         
   │                                                                                                                  
   │             ┌──────────────────────────────────────────             ──────────────────────────────┐┌ ─ ─ ─ ─ ─ ─ 
   │  Compaction │                    RUB                        ...                                   │     MUB     │
   │      N      └──────────────────────────────────────────             ──────────────────────────────┘└ ─ ─ ─ ─ ─ ─ 
   │                                                                                                                  
   │                                                                                                                  
   │                                                                                                                  
   │                                                                                                                  
   │                                                                                                                  
   │                                                                                                                  
   ▼                                                                                                                  
   Wallclock
```

The current idea is, during compacting, split RUB into 2 RUBs: a large one that is not overlapped with MUB and a small one that is potentially overlapped with MUB. Next round of compacting only needs to compact the small one with the new MUB.

**Comment by pauldix** [comment]
This split must happen at some point anyway for persistence to have the system persist non-overlapping Parquet files. Some amount of write amplification in memory due to compaction is unavoidable. In the latest data lifecycle doc we aimed for some number of compactions until we'd get to a RUB chunk that could be split.

We have the MUB row threshold, after which a MUB chunk should get compacted into the RUB. We also have the RUB persist threshold, after which, the RUB chunk will be persisted and then no longer be a candidate for compaction. The current defaults for these are 100k and 1M respectively, which would mean only 9 compactions (1 create and 9 compact) before a RUB chunk would move over to being persisted.

However, the last step in this transition is take the MUB + RUB and split on a timestamp from the persist windows, leaving us with a large RUB chunk that will be non-overlapping with smaller RUB chunk that will be overlapping with whatever is arriving in the MUB.

In practice, the MUB is always going to be overlapping with the leading edge chunk in the RUB. At least until a partition goes cold for writes.

We could avoid some amount of amplification by having levels, but what makes this system strange is that the level sort order and key space sort order are different. Our levels need to be based strictly on time (i.e. we want blocks of data that don't overlap in the time column). But the data within a level is sorted by tagset low cardinality to high and then time. I had documented this in some previous lifecycle doc, which is what led to this simpler algorithm.

The window of time for ingesting overlapping data is a function of clock skew in the systems assigning timestamp and lags in data collection. For example, if we have 1 minute of potential lag, we need to buffer up a few minutes worth of data before splitting it out into a block that won't overlap with data that is going to be arriving with the rest.

We could build up some of these smaller RUB chunks via this split mechanism and then compact those together later. Although it's not yet clear to me that we need that complexity. Particularly if items #1 and #2 from #1885 are addressed, which needs to happen anyway for query performance.

Anyway, if we get the split working for persisting RUB chunks, we'll get this property without building out a more complex compaction algorithm.
